#Loading the data into specific dataframes

In [239]:
import pandas as pd

In [240]:
# Mount your Google Drive to Colab
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


##Order Report

In [241]:
# Specify the path of your Excel file in Google Drive
file_path = '/content/drive/MyDrive/Cointab-Assignment/Company X - Order Report.xlsx'

In [242]:
# Read the Excel file into a pandas dataframe
X_order_report = pd.read_excel(file_path)
print(X_order_report.shape)

(400, 3)


In [243]:
X_order_report.columns

Index(['ExternOrderNo', 'SKU', 'Order Qty'], dtype='object')

The table contains 400 order details including order number, SKU and Order quantity

In [244]:
X_order_report.head()

ExternOrderNo            SKU  Order Qty
0     2001827036  8904223818706        1.0
1     2001827036  8904223819093        1.0
2     2001827036  8904223819109        1.0
3     2001827036  8904223818430        1.0
4     2001827036  8904223819277        1.0

In [245]:
X_order_report.tail()

ExternOrderNo            SKU  Order Qty
395     2001806229  8904223818942        1.0
396     2001806229  8904223818850        1.0
397     2001806226  8904223818850        2.0
398     2001806210  8904223816214        1.0
399     2001806210  8904223818874        1.0

In [246]:
duplicates = X_order_report[X_order_report.duplicated()]

print(duplicates)

     ExternOrderNo            SKU  Order Qty
110     2001811363  8904223815859        1.0
317     2001806885  8904223819499        2.0


Two records are found as duplicates

In [247]:
#Dropping the duplicated records

X_order_report = X_order_report.drop_duplicates()

print(f"The new shape of the order report is {X_order_report.shape}.")

The new shape of the order report is (398, 3).


In [248]:
#Finding null values
X_order_report.isnull().sum()

ExternOrderNo    0
SKU              0
Order Qty        0
dtype: int64

No null values found

In [249]:
#changing the column orderId name
X_order_report.columns = ["Order ID", "SKU", "Order Qty"]

In [250]:
X_order_report.sample()

Order ID            SKU  Order Qty
70  2001814580  8904223818706        1.0

In [251]:
X_order_report.dtypes

Order ID       int64
SKU           object
Order Qty    float64
dtype: object

##Pincode Zones

In [252]:
# Specify the path of your Excel file in Google Drive
file_path = '/content/drive/MyDrive/Cointab-Assignment/Company X - Pincode Zones.xlsx'

In [253]:
# Read the Excel file into a pandas dataframe
Pincode_map = pd.read_excel(file_path)
print(Pincode_map.shape)

(124, 3)


In [254]:
Pincode_map.columns

Index(['Warehouse Pincode', 'Customer Pincode', 'Zone'], dtype='object')

The table contains 124 order details including 'Warehouse Pincode', 'Customer Pincode', 'Zone'

In [255]:
Pincode_map.head()

Warehouse Pincode  Customer Pincode Zone
0           121003.0          507101.0    d
1           121003.0          486886.0    d
2           121003.0          532484.0    d
3           121003.0          143001.0    b
4           121003.0          515591.0    d

In [256]:
Pincode_map.tail()

Warehouse Pincode  Customer Pincode Zone
119           121003.0          325207.0    b
120           121003.0          303702.0    b
121           121003.0          313301.0    b
122           121003.0          173212.0    e
123           121003.0          302020.0    b

In [257]:
duplicates = Pincode_map[Pincode_map.duplicated()]

print(duplicates)

     Warehouse Pincode  Customer Pincode Zone
16            121003.0          140301.0    b
50            121003.0          248001.0    b
72            121003.0          302002.0    b
75            121003.0          302017.0    b
77            121003.0          313001.0    b
78            121003.0          313001.0    b
82            121003.0          313001.0    b
90            121003.0          302017.0    b
101           121003.0          335001.0    b
106           121003.0          302031.0    b
113           121003.0          302017.0    b
116           121003.0          302018.0    b
117           121003.0          302017.0    b
118           121003.0          302012.0    b
122           121003.0          173212.0    e
123           121003.0          302020.0    b


16 records are found as duplicates

In [258]:
#Dropping the duplicated records

Pincode_map = Pincode_map.drop_duplicates()

print(f"The new shape of the Pincode zone is {Pincode_map.shape}.")

The new shape of the Pincode zone is (108, 3).


In [259]:
#Finding null values
Pincode_map.isnull().sum()

Warehouse Pincode    0
Customer Pincode     0
Zone                 0
dtype: int64

No null values found

##SKU Master

In [260]:
# Specify the path of your Excel file in Google Drive
file_path = '/content/drive/MyDrive/Cointab-Assignment/Company X - SKU Master.xlsx'

In [261]:
# Read the Excel file into a pandas dataframe
SKU_Master = pd.read_excel(file_path)
print(SKU_Master.shape)

(66, 2)


In [262]:
SKU_Master.columns

Index(['SKU', 'Weight (g)'], dtype='object')

The table contains 66 order details including 'SKU', 'Weight (g)'

In [263]:
SKU_Master.head()

SKU  Weight (g)
0  8904223815682         210
1  8904223815859         165
2  8904223815866         113
3  8904223815873          65
4  8904223816214         120

In [264]:
SKU_Master.tail()

SKU  Weight (g)
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

In [265]:
duplicates = SKU_Master[SKU_Master.duplicated()]

print(duplicates)

              SKU  Weight (g)
56  GIFTBOX202002         500


One record found as duplicates

In [266]:
#Dropping the duplicated records

SKU_Master = SKU_Master.drop_duplicates()

print(f"The new shape of the SKU Master data is {SKU_Master.shape}.")

The new shape of the SKU Master data is (65, 2).


In [267]:
#Finding null values
SKU_Master.isnull().sum()

SKU           0
Weight (g)    0
dtype: int64

No null values found

##Courier Company - Rates

In [268]:
# Specify the path of your Excel file in Google Drive
file_path = '/content/drive/MyDrive/Cointab-Assignment/Courier Company - Rates.xlsx'

In [269]:
# Read the Excel file into a pandas dataframe
Courier_rate = pd.read_excel(file_path)
print(Courier_rate.shape)

(5, 6)


In [270]:
Courier_rate.columns

Index(['Zone', 'Weight Slabs', 'Forward Fixed Charge',
       'Forward Additional Weight Slab Charge', 'RTO Fixed Charge',
       'RTO Additional Weight Slab Charge'],
      dtype='object')

The table contains 5 order details including 'Zone', 'Weight Slabs', 'Forward Fixed Charge',
       'Forward Additional Weight Slab Charge', 'RTO Fixed Charge',
       'RTO Additional Weight Slab Charge'

In [271]:
Courier_rate.head()

Zone  Weight Slabs  Forward Fixed Charge  \
0    A          0.25                  29.5   
1    B          0.50                  33.0   
2    C          0.75                  40.1   
3    D          1.25                  45.4   
4    E          1.50                  56.6   

   Forward Additional Weight Slab Charge  RTO Fixed Charge  \
0                                   23.6              13.6   
1                                   28.3              20.5   
2                                   38.9              31.9   
3                                   44.8              41.3   
4                                   55.5              50.7   

   RTO Additional Weight Slab Charge  
0                               23.6  
1                               28.3  
2                               38.9  
3                               44.8  
4                               55.5

In [272]:
Courier_rate.tail()

Zone  Weight Slabs  Forward Fixed Charge  \
0    A          0.25                  29.5   
1    B          0.50                  33.0   
2    C          0.75                  40.1   
3    D          1.25                  45.4   
4    E          1.50                  56.6   

   Forward Additional Weight Slab Charge  RTO Fixed Charge  \
0                                   23.6              13.6   
1                                   28.3              20.5   
2                                   38.9              31.9   
3                                   44.8              41.3   
4                                   55.5              50.7   

   RTO Additional Weight Slab Charge  
0                               23.6  
1                               28.3  
2                               38.9  
3                               44.8  
4                               55.5

In [273]:
duplicates = Courier_rate[Courier_rate.duplicated()]

print(duplicates)

Empty DataFrame
Columns: [Zone, Weight Slabs, Forward Fixed Charge, Forward Additional Weight Slab Charge, RTO Fixed Charge, RTO Additional Weight Slab Charge]
Index: []


No duplicate records found

In [274]:
#Finding null values
Courier_rate.isnull().sum()

Zone                                     0
Weight Slabs                             0
Forward Fixed Charge                     0
Forward Additional Weight Slab Charge    0
RTO Fixed Charge                         0
RTO Additional Weight Slab Charge        0
dtype: int64

No null values found

##Courier Company - Invoice

In [275]:
# Specify the path of your Excel file in Google Drive
file_path = '/content/drive/MyDrive/Cointab-Assignment/Courier Company - Invoice.xlsx'

In [276]:
# Read the Excel file into a pandas dataframe
Courier_invoice = pd.read_excel(file_path)
print(Courier_invoice.shape)

(124, 8)


In [277]:
Courier_invoice.columns

Index(['AWB Code', 'Order ID', 'Charged Weight', 'Warehouse Pincode',
       'Customer Pincode', 'Zone', 'Type of Shipment', 'Billing Amount (Rs.)'],
      dtype='object')

The table contains 124 order details including 'AWB Code', 'Order ID', 'Charged Weight', 'Warehouse Pincode',
       'Customer Pincode', 'Zone', 'Type of Shipment', 'Billing Amount (Rs.)'

In [278]:
Courier_invoice.head()

AWB Code      Order ID  Charged Weight  Warehouse Pincode  \
0  1091117222124  2.001806e+09            1.30           121003.0   
1  1091117222194  2.001806e+09            1.00           121003.0   
2  1091117222931  2.001806e+09            2.50           121003.0   
3  1091117223244  2.001806e+09            1.00           121003.0   
4  1091117229345  2.001807e+09            0.15           121003.0   

   Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
0            507101    d  Forward charges                 135.0  
1            486886    d  Forward charges                  90.2  
2            532484    d  Forward charges                 224.6  
3            143001    b  Forward charges                  61.3  
4            515591    d  Forward charges                  45.4

In [279]:
Courier_invoice.tail()

AWB Code      Order ID  Charged Weight  Warehouse Pincode  \
119  1091118551656  2.001813e+09            0.73           121003.0   
120  1091117614452  2.001809e+09            0.50           121003.0   
121  1091120922803  2.001821e+09            0.50           121003.0   
122  1091121844806  2.001811e+09            0.50           121003.0   
123  1091121846136  2.001811e+09            0.50           121003.0   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
119            325207    d          Forward charges                  90.2  
120            303702    d  Forward and RTO charges                  86.7  
121            313301    d          Forward charges                  45.4  
122            173212    b          Forward charges                  33.0  
123            302020    d          Forward charges                  45.4

In [280]:
Courier_invoice["Order ID"].nunique()

124

In [281]:
duplicates = Courier_invoice[Courier_invoice.duplicated()]

print(duplicates)

Empty DataFrame
Columns: [AWB Code, Order ID, Charged Weight, Warehouse Pincode, Customer Pincode, Zone, Type of Shipment, Billing Amount (Rs.)]
Index: []


No duplicate records found

In [282]:
#Finding null values
Courier_invoice.isnull().sum()

AWB Code                0
Order ID                0
Charged Weight          0
Warehouse Pincode       0
Customer Pincode        0
Zone                    0
Type of Shipment        0
Billing Amount (Rs.)    0
dtype: int64

No null values found

In [283]:
Courier_invoice["Warehouse Pincode"].value_counts()

121003.0    124
Name: Warehouse Pincode, dtype: int64

In [284]:
Courier_invoice["Customer Pincode"].value_counts()

302017    5
313001    4
302002    2
302031    2
248001    2
         ..
723146    1
140604    1
831006    1
562110    1
313301    1
Name: Customer Pincode, Length: 108, dtype: int64

In [285]:
Courier_invoice["Type of Shipment"].value_counts()

Forward charges            109
Forward and RTO charges     15
Name: Type of Shipment, dtype: int64

In [286]:
Courier_invoice.dtypes

AWB Code                  int64
Order ID                float64
Charged Weight          float64
Warehouse Pincode       float64
Customer Pincode          int64
Zone                     object
Type of Shipment         object
Billing Amount (Rs.)    float64
dtype: object

##Merging columns

In [287]:
merged_table = pd.merge(X_order_report, SKU_Master, how = "inner")
merged_table.shape

(398, 4)

In [288]:
merged_table.tail()

Order ID            SKU  Order Qty  Weight (g)
393  2001806616  8904223819123        1.0         250
394  2001806567  8904223815804        1.0         160
395  2001806567  8904223818577        1.0         150
396  2001806408  8904223819437        2.0         552
397  2001806232  8904223818645        6.0         137

In [289]:
merged_table["Weight (g)"] = merged_table["Weight (g)"] * merged_table["Order Qty"]
merged_table.tail(10)

Order ID            SKU  Order Qty  Weight (g)
388  2001807976  8904223818713        1.0       120.0
389  2001807415  8904223818713        1.0       120.0
390  2001807058  8904223818713        1.0       120.0
391  2001806616  8904223818713        1.0       120.0
392  2001807328  8904223818997        1.0       490.0
393  2001806616  8904223819123        1.0       250.0
394  2001806567  8904223815804        1.0       160.0
395  2001806567  8904223818577        1.0       150.0
396  2001806408  8904223819437        2.0      1104.0
397  2001806232  8904223818645        6.0       822.0

In [290]:
merged_table = merged_table.groupby(merged_table["Order ID"]).agg({"Weight (g)": "sum"}).reset_index()
merged_table.shape

(124, 2)

In [291]:
merged_table.head()

Order ID  Weight (g)
0  2001806210       220.0
1  2001806226       480.0
2  2001806229       500.0
3  2001806232      1302.0
4  2001806233       245.0

In [292]:
# Perform VLOOKUP-like operation
merged_table = merged_table.merge(Courier_invoice, how='inner')

# Print the merged table
merged_table.head()

Order ID  Weight (g)       AWB Code  Charged Weight  Warehouse Pincode  \
0  2001806210       220.0  1091117221940            2.92           121003.0   
1  2001806226       480.0  1091117222065            0.68           121003.0   
2  2001806229       500.0  1091117222080            0.71           121003.0   
3  2001806232      1302.0  1091117222124            1.30           121003.0   
4  2001806233       245.0  1091117222135            0.78           121003.0   

   Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
0            140604    b  Forward charges                 174.5  
1            723146    d  Forward charges                  90.2  
2            421204    d  Forward charges                  90.2  
3            507101    d  Forward charges                 135.0  
4            263139    b  Forward charges                  61.3

In [293]:
merged_table.tail()

Order ID  Weight (g)       AWB Code  Charged Weight  Warehouse Pincode  \
119  2001821995       477.0  1091121183730             0.5           121003.0   
120  2001822466      1376.0  1091121305541             1.1           121003.0   
121  2001823564       672.0  1091121666133             0.7           121003.0   
122  2001825261      1557.0  1091121981575             1.6           121003.0   
123  2001827036      1676.0  1091122418320             1.6           121003.0   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
119            342008    d          Forward charges                  45.4  
120            342301    d          Forward charges                 135.0  
121            492001    d  Forward and RTO charges                 172.8  
122            517128    d  Forward and RTO charges                 345.0  
123            173213    b          Forward charges                 117.9

In [294]:
duplicates  = merged_table[merged_table.duplicated()]

print(duplicates)

Empty DataFrame
Columns: [Order ID, Weight (g), AWB Code, Charged Weight, Warehouse Pincode, Customer Pincode, Zone, Type of Shipment, Billing Amount (Rs.)]
Index: []


In [295]:
merged_table.isnull().sum()

Order ID                0
Weight (g)              0
AWB Code                0
Charged Weight          0
Warehouse Pincode       0
Customer Pincode        0
Zone                    0
Type of Shipment        0
Billing Amount (Rs.)    0
dtype: int64

In [296]:
merged_table = merged_table.merge(Pincode_map, right_on = "Customer Pincode", left_on= "Customer Pincode")
merged_table.sample(2)

Order ID  Weight (g)       AWB Code  Charged Weight  \
29   2001806776       611.0  1091117227573            2.86   
114  2001821190      1200.0  1091120959015            1.20   

     Warehouse Pincode_x  Customer Pincode Zone_x         Type of Shipment  \
29              121003.0            226004      b          Forward charges   
114             121003.0            486661      d  Forward and RTO charges   

     Billing Amount (Rs.)  Warehouse Pincode_y Zone_y  
29                  174.5             121003.0      b  
114                 258.9             121003.0      d

In [297]:
merged_table = merged_table.drop(["Warehouse Pincode_x", "Customer Pincode", "Warehouse Pincode_y"], axis = 1)

In [298]:
merged_table.shape

(124, 8)

In [299]:
merged_table.sample()

Order ID  Weight (g)       AWB Code  Charged Weight Zone_x  \
19  2001806567       952.0  1091117224902            1.16      d   

   Type of Shipment  Billing Amount (Rs.) Zone_y  
19  Forward charges                 135.0      d

In [300]:
merged_table.columns = ['Order ID', 'Total Weight (g)', 'AWB Code', 'C_Charged Weight', 'C_Zone',
       'Type of Shipment', 'C_Billing Amount (Rs.)', 'X_Zone']

In [301]:
merged_table.sample()

Order ID  Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
82  2001810104             601.0  1091117804200              0.76      d   

   Type of Shipment  C_Billing Amount (Rs.) X_Zone  
82  Forward charges                    90.2      b

In [302]:
Courier_rate["Zone"] = Courier_rate["Zone"].str.lower()

In [303]:
Courier_rate.head()

Zone  Weight Slabs  Forward Fixed Charge  \
0    a          0.25                  29.5   
1    b          0.50                  33.0   
2    c          0.75                  40.1   
3    d          1.25                  45.4   
4    e          1.50                  56.6   

   Forward Additional Weight Slab Charge  RTO Fixed Charge  \
0                                   23.6              13.6   
1                                   28.3              20.5   
2                                   38.9              31.9   
3                                   44.8              41.3   
4                                   55.5              50.7   

   RTO Additional Weight Slab Charge  
0                               23.6  
1                               28.3  
2                               38.9  
3                               44.8  
4                               55.5

In [304]:
merged_table = merged_table.merge(Courier_rate, left_on="C_Zone", right_on="Zone")
merged_table.shape

(124, 14)

In [305]:
merged_table.head()

Order ID  Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
0  2001806210             220.0  1091117221940              2.92      b   
1  2001806233             245.0  1091117222135              0.78      b   
2  2001806458             700.0  1091117223244              1.00      b   
3  2001806547             127.0  1091117224611              1.00      b   
4  2001806575             500.0  1091117225016              0.68      b   

  Type of Shipment  C_Billing Amount (Rs.) X_Zone Zone  Weight Slabs  \
0  Forward charges                   174.5      b    b           0.5   
1  Forward charges                    61.3      b    b           0.5   
2  Forward charges                    61.3      b    b           0.5   
3  Forward charges                    61.3      b    b           0.5   
4  Forward charges                    61.3      b    b           0.5   

   Forward Fixed Charge  Forward Additional Weight Slab Charge  \
0                  33.0                                   28.3   
1                  33.0                                   28.3   
2                  33.0                                   28.3   
3                  33.0                                   28.3   
4                  33.0                                   28.3   

   RTO Fixed Charge  RTO Additional Weight Slab Charge  
0              20.5                               28.3  
1              20.5                               28.3  
2              20.5                               28.3  
3              20.5                               28.3  
4              20.5                               28.3

In [306]:
merged_table = merged_table.drop("Zone", axis = 1)
merged_table.head()

Order ID  Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
0  2001806210             220.0  1091117221940              2.92      b   
1  2001806233             245.0  1091117222135              0.78      b   
2  2001806458             700.0  1091117223244              1.00      b   
3  2001806547             127.0  1091117224611              1.00      b   
4  2001806575             500.0  1091117225016              0.68      b   

  Type of Shipment  C_Billing Amount (Rs.) X_Zone  Weight Slabs  \
0  Forward charges                   174.5      b           0.5   
1  Forward charges                    61.3      b           0.5   
2  Forward charges                    61.3      b           0.5   
3  Forward charges                    61.3      b           0.5   
4  Forward charges                    61.3      b           0.5   

   Forward Fixed Charge  Forward Additional Weight Slab Charge  \
0                  33.0                                   28.3   
1                  33.0                                   28.3   
2                  33.0                                   28.3   
3                  33.0                                   28.3   
4                  33.0                                   28.3   

   RTO Fixed Charge  RTO Additional Weight Slab Charge  
0              20.5                               28.3  
1              20.5                               28.3  
2              20.5                               28.3  
3              20.5                               28.3  
4              20.5                               28.3

In [307]:
merged_table.columns = ['Order ID', 'X_Total Weight (g)', 'AWB Code', 'C_Charged Weight',
       'C_Zone', 'Type of Shipment', 'C_Billing Amount (Rs.)', 'X_Zone',
       'C_Weight Slabs', 'C_Forward Fixed Charge',
       'C_Forward Additional Weight Slab Charge', 'C_RTO Fixed Charge',
       'C_RTO Additional Weight Slab Charge']

In [308]:
merged_table.sample()

Order ID  X_Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
11  2001807415              1048.0  1091117325094               1.0      b   

   Type of Shipment  C_Billing Amount (Rs.) X_Zone  C_Weight Slabs  \
11  Forward charges                    61.3      b             0.5   

    C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
11                    33.0                                     28.3   

    C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  
11                20.5                                 28.3

In [309]:
merged_table["verify_zone"] = (merged_table["X_Zone"]==merged_table["C_Zone"]).astype(int)
merged_table["verify_zone"].value_counts()

0    65
1    59
Name: verify_zone, dtype: int64

Almost 50% of delivery zones are different between company X and courier company

In [310]:
merged_table.head()

Order ID  X_Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
0  2001806210               220.0  1091117221940              2.92      b   
1  2001806233               245.0  1091117222135              0.78      b   
2  2001806458               700.0  1091117223244              1.00      b   
3  2001806547               127.0  1091117224611              1.00      b   
4  2001806575               500.0  1091117225016              0.68      b   

  Type of Shipment  C_Billing Amount (Rs.) X_Zone  C_Weight Slabs  \
0  Forward charges                   174.5      b             0.5   
1  Forward charges                    61.3      b             0.5   
2  Forward charges                    61.3      b             0.5   
3  Forward charges                    61.3      b             0.5   
4  Forward charges                    61.3      b             0.5   

   C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
0                    33.0                                     28.3   
1                    33.0                                     28.3   
2                    33.0                                     28.3   
3                    33.0                                     28.3   
4                    33.0                                     28.3   

   C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  verify_zone  
0                20.5                                 28.3            1  
1                20.5                                 28.3            1  
2                20.5                                 28.3            1  
3                20.5                                 28.3            1  
4                20.5                                 28.3            1

In [311]:
merged_table = merged_table.merge(Courier_rate, left_on="X_Zone", right_on="Zone")
merged_table.shape

(124, 20)

In [312]:
merged_table.sample()

Order ID  X_Total Weight (g)       AWB Code  C_Charged Weight C_Zone  \
89  2001806251               245.0  1091117222146              1.27      d   

   Type of Shipment  C_Billing Amount (Rs.) X_Zone  C_Weight Slabs  \
89  Forward charges                   135.0      d            1.25   

    C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
89                    45.4                                     44.8   

    C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  verify_zone Zone  \
89                41.3                                 44.8            1    d   

    Weight Slabs  Forward Fixed Charge  Forward Additional Weight Slab Charge  \
89          1.25                  45.4                                   44.8   

    RTO Fixed Charge  RTO Additional Weight Slab Charge  
89              41.3                               44.8

In [313]:
merged_table = merged_table.drop(['verify_zone', 'Zone', 'Forward Fixed Charge',
       'Forward Additional Weight Slab Charge', 'RTO Fixed Charge',
       'RTO Additional Weight Slab Charge'], axis = 1)

In [314]:
merged_table.columns = ['Order ID', 'X_Total Weight (g)', 'AWB Code', 'C_Charged_Weight',
       'C_Zone', 'C_Type of Shipment', 'C_Billing_Amount(Rs.', 'X_Zone',
       'C_Weight Slabs', 'C_Forward Fixed Charge',
       'C_Forward Additional Weight Slab Charge', 'C_RTO Fixed Charge',
       'C_RTO Additional Weight Slab Charge', 'X_Weight Slabs']

In [315]:
merged_table.shape

(124, 14)

In [316]:
merged_table.sample()

Order ID  X_Total Weight (g)       AWB Code  C_Charged_Weight C_Zone  \
12  2001807814               607.0  1091117326612              0.79      b   

   C_Type of Shipment  C_Billing_Amount(Rs. X_Zone  C_Weight Slabs  \
12    Forward charges                  61.3      b             0.5   

    C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
12                    33.0                                     28.3   

    C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  X_Weight Slabs  
12                20.5                                 28.3             0.5

In [317]:
rearrange = ['Order ID', 'AWB Code', 'X_Total Weight (g)', 'X_Weight Slabs', 'C_Charged_Weight', 
             'C_Weight Slabs',  'X_Zone', 'C_Zone', 'C_Billing_Amount(Rs.',
             'C_Forward Fixed Charge', 'C_Forward Additional Weight Slab Charge',
             'C_RTO Fixed Charge', 'C_RTO Additional Weight Slab Charge', "C_Type of Shipment"]

In [318]:
Output = merged_table

In [319]:
Output = Output.reindex(columns = rearrange)
Output.sample()

Order ID       AWB Code  X_Total Weight (g)  X_Weight Slabs  \
62  2001811809  1091118009786               500.0             0.5   

    C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
62               0.5            1.25      b      d                  86.7   

    C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
62                    45.4                                     44.8   

    C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  \
62                41.3                                 44.8   

         C_Type of Shipment  
62  Forward and RTO charges

In [320]:
Output.shape

(124, 14)

In [321]:
# Converting unit of weight from grams to KG
Output["X_Total Weight (g)"] = round(Output["X_Total Weight (g)"]/1000,2)

Output.columns = ['Order ID', 'AWB Code', 'X_Total Weight', 'X_Weight Slabs', 'C_Charged_Weight', 
             'C_Weight Slabs',  'X_Zone', 'C_Zone', 'C_Billing_Amount(Rs.',
             'C_Forward Fixed Charge', 'C_Forward Additional Weight Slab Charge',
             'C_RTO Fixed Charge', 'C_RTO Additional Weight Slab Charge', "C_Type of Shipment"]
Output.head()

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
0  2001806210  1091117221940            0.22             0.5   
1  2001806233  1091117222135            0.24             0.5   
2  2001806458  1091117223244            0.70             0.5   
3  2001806547  1091117224611            0.13             0.5   
4  2001806575  1091117225016            0.50             0.5   

   C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
0              2.92             0.5      b      b                 174.5   
1              0.78             0.5      b      b                  61.3   
2              1.00             0.5      b      b                  61.3   
3              1.00             0.5      b      b                  61.3   
4              0.68             0.5      b      b                  61.3   

   C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
0                    33.0                                     28.3   
1                    33.0                                     28.3   
2                    33.0                                     28.3   
3                    33.0                                     28.3   
4                    33.0                                     28.3   

   C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge C_Type of Shipment  
0                20.5                                 28.3    Forward charges  
1                20.5                                 28.3    Forward charges  
2                20.5                                 28.3    Forward charges  
3                20.5                                 28.3    Forward charges  
4                20.5                                 28.3    Forward charges

In [322]:
def calculate_applicable_weight(Total_weight, Weight_slab):
  if Total_weight > Weight_slab:
    return Weight_slab * (int(Total_weight/Weight_slab) + 1)
  else:
    return Weight_slab

In [323]:
Output["Applicable_weight"] = Output.apply(lambda x: calculate_applicable_weight(x["X_Total Weight"], x["X_Weight Slabs"]), axis = 1)

In [324]:
Output.tail()

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
119  2001816684  1091119398844            0.92            1.25   
120  2001817160  1091119630264            0.70            1.25   
121  2001821190  1091120959015            1.20            1.25   
122  2001823564  1091121666133            0.67            1.25   
123  2001825261  1091121981575            1.56            1.25   

     C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
119              0.99            1.25      d      d                 172.8   
120              0.70            1.25      d      d                 172.8   
121              1.20            1.25      d      d                 258.9   
122              0.70            1.25      d      d                 172.8   
123              1.60            1.25      d      d                 345.0   

     C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
119                    45.4                                     44.8   
120                    45.4                                     44.8   
121                    45.4                                     44.8   
122                    45.4                                     44.8   
123                    45.4                                     44.8   

     C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  \
119                41.3                                 44.8   
120                41.3                                 44.8   
121                41.3                                 44.8   
122                41.3                                 44.8   
123                41.3                                 44.8   

          C_Type of Shipment  Applicable_weight  
119  Forward and RTO charges               1.25  
120  Forward and RTO charges               1.25  
121  Forward and RTO charges               1.25  
122  Forward and RTO charges               1.25  
123  Forward and RTO charges               2.50

In [325]:
#Changing the values of RTO charges to zero for the orders that have shipment type of "Forward charges Only"
Output[['C_RTO Fixed Charge', 'C_RTO Additional Weight Slab Charge']] = [(0,0) if "RTO" not in row["C_Type of Shipment"] 
                                                                         else (row["C_RTO Fixed Charge"], row['C_RTO Additional Weight Slab Charge'])
                                                                               for idx, row in Output.iterrows()]

In [326]:
Output.sample(10)

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
71   2001820690  1091121306101            0.06            0.50   
43   2001808832  1091117437864            0.60            0.50   
42   2001808739  1091117437293            0.22            0.50   
19   2001806304  1091117222360            0.50            0.50   
20   2001810549  1091117806263            0.99            0.50   
123  2001825261  1091121981575            1.56            1.25   
90   2001806273  1091117222194            0.62            1.25   
4    2001806575  1091117225016            0.50            0.50   
60   2001811466  1091118001865            0.61            0.50   
95   2001806567  1091117224902            0.95            1.25   

     C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
71               0.15            1.25      b      d                  45.4   
43               2.47            1.25      b      d                 224.6   
42               1.63            1.25      b      d                 179.8   
19               0.71            1.25      b      d                  90.2   
20               1.86            1.25      b      d                 179.8   
123              1.60            1.25      d      d                 345.0   
90               1.00            1.25      d      d                  90.2   
4                0.68            0.50      b      b                  61.3   
60               0.80            1.25      b      d                  90.2   
95               1.16            1.25      d      d                 135.0   

     C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
71                     45.4                                     44.8   
43                     45.4                                     44.8   
42                     45.4                                     44.8   
19                     45.4                                     44.8   
20                     45.4                                     44.8   
123                    45.4                                     44.8   
90                     45.4                                     44.8   
4                      33.0                                     28.3   
60                     45.4                                     44.8   
95                     45.4                                     44.8   

     C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  \
71                  0.0                                  0.0   
43                  0.0                                  0.0   
42                  0.0                                  0.0   
19                  0.0                                  0.0   
20                  0.0                                  0.0   
123                41.3                                 44.8   
90                  0.0                                  0.0   
4                   0.0                                  0.0   
60                  0.0                                  0.0   
95                  0.0                                  0.0   

          C_Type of Shipment  Applicable_weight  
71           Forward charges               0.50  
43           Forward charges               1.00  
42           Forward charges               0.50  
19           Forward charges               0.50  
20           Forward charges               1.00  
123  Forward and RTO charges               2.50  
90           Forward charges               1.25  
4            Forward charges               0.50  
60           Forward charges               1.00  
95           Forward charges               1.25

In [327]:
Output = Output.drop("C_Type of Shipment", axis= 1)

In [328]:
Output.sample()

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
9  2001807058  1091117323112            1.17             0.5   

   C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
9              1.15             0.5      b      b                  89.6   

   C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
9                    33.0                                     28.3   

   C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  Applicable_weight  
9                 0.0                                  0.0                1.5

In [329]:
def calculate_charge(weight, weight_slab, fixed_fwd, fwd_additional, fixed_rto, rto_additional):
    n = int(weight / weight_slab)  # calculate the number of slabs
    if weight <= weight_slab:
        charge = fixed_fwd + fixed_rto
    else:
        charge = fixed_fwd + (n - 1) * fwd_additional + fixed_rto + (n - 1) * rto_additional
    return charge


In [330]:
Output['Expected_charge'] = Output.apply(lambda x: calculate_charge(x['Applicable_weight'], x['X_Weight Slabs'], 
                                                                    x["C_Forward Fixed Charge"], 
                                                                      x["C_Forward Additional Weight Slab Charge"], 
                                                                      x["C_RTO Fixed Charge"],	x["C_RTO Additional Weight Slab Charge"]),
                                          axis=1)


In [331]:
Output.sample(10)

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
59   2001811363  1091117958395            0.34            0.50   
24   2001806471  1091117223351            1.62            0.50   
1    2001806233  1091117222135            0.24            0.50   
34   2001807852  1091117326925            0.50            0.50   
60   2001811466  1091118001865            0.61            0.50   
104  2001807084  1091117323215            0.50            1.25   
42   2001808739  1091117437293            0.22            0.50   
78   2001822466  1091121305541            1.38            0.50   
43   2001808832  1091117437864            0.60            0.50   
30   2001807328  1091117324346            0.49            0.50   

     C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
59               0.59            1.25      b      d                  90.2   
24               1.70            1.25      b      d                 179.8   
1                0.78            0.50      b      b                  61.3   
34               0.74            1.25      b      d                  90.2   
60               0.80            1.25      b      d                  90.2   
104              0.67            1.25      d      d                  90.2   
42               1.63            1.25      b      d                 179.8   
78               1.10            1.25      b      d                 135.0   
43               2.47            1.25      b      d                 224.6   
30               2.28            1.25      b      d                 224.6   

     C_Forward Fixed Charge  C_Forward Additional Weight Slab Charge  \
59                     45.4                                     44.8   
24                     45.4                                     44.8   
1                      33.0                                     28.3   
34                     45.4                                     44.8   
60                     45.4                                     44.8   
104                    45.4                                     44.8   
42                     45.4                                     44.8   
78                     45.4                                     44.8   
43                     45.4                                     44.8   
30                     45.4                                     44.8   

     C_RTO Fixed Charge  C_RTO Additional Weight Slab Charge  \
59                  0.0                                  0.0   
24                  0.0                                  0.0   
1                   0.0                                  0.0   
34                  0.0                                  0.0   
60                  0.0                                  0.0   
104                 0.0                                  0.0   
42                  0.0                                  0.0   
78                  0.0                                  0.0   
43                  0.0                                  0.0   
30                  0.0                                  0.0   

     Applicable_weight  Expected_charge  
59                0.50             45.4  
24                2.00            179.8  
1                 0.50             33.0  
34                0.50             45.4  
60                1.00             90.2  
104               1.25             45.4  
42                0.50             45.4  
78                1.50            135.0  
43                1.00             90.2  
30                0.50             45.4

In [332]:
rearrange = ['Order ID', 'AWB Code', 'X_Total Weight', 'X_Weight Slabs',
             'C_Charged_Weight', 'C_Weight Slabs', 'X_Zone', 'C_Zone', 
             'C_Forward Fixed Charge', 'C_Forward Additional Weight Slab Charge', 
             'C_RTO Fixed Charge', 'C_RTO Additional Weight Slab Charge', 
             'Applicable_weight', 'C_Billing_Amount(Rs.', 'Expected_charge']

In [333]:
Output = Output.reindex(columns=rearrange)
Output.sample()

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
93  2001806446  1091117223211             0.5            1.25   

    C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Forward Fixed Charge  \
93              0.69            1.25      d      d                    45.4   

    C_Forward Additional Weight Slab Charge  C_RTO Fixed Charge  \
93                                     44.8                 0.0   

    C_RTO Additional Weight Slab Charge  Applicable_weight  \
93                                  0.0               1.25   

    C_Billing_Amount(Rs.  Expected_charge  
93                  90.2             45.4

In [334]:
Output = Output.drop(['C_Forward Fixed Charge', 'C_Forward Additional Weight Slab Charge',
             'C_RTO Fixed Charge', 'C_RTO Additional Weight Slab Charge', 'Applicable_weight'], axis = 1)

In [335]:
Output["Difference(Courier-Expected_X)"] = round(Output["C_Billing_Amount(Rs."]-Output["Expected_charge"],1)
Output.sample(10)

Order ID       AWB Code  X_Total Weight  X_Weight Slabs  \
102  2001807012  1091117229345            0.24            1.25   
56   2001811305  1091121846136            0.75            0.50   
79   2001808585  1091117436652            0.50            1.50   
80   2001811475  1091121844806            0.69            1.50   
85   2001818390  1091120014461            0.84            1.50   
43   2001808832  1091117437864            0.60            0.50   
54   2001810697  1091117807140            0.61            0.50   
103  2001807036  1091117323005            1.46            1.25   
2    2001806458  1091117223244            0.70            0.50   
3    2001806547  1091117224611            0.13            0.50   

     C_Charged_Weight  C_Weight Slabs X_Zone C_Zone  C_Billing_Amount(Rs.  \
102              0.15            1.25      d      d                  45.4   
56               0.50            1.25      b      d                  45.4   
79               0.72            0.50      e      b                  61.3   
80               0.50            0.50      e      b                  33.0   
85               0.80            1.50      e      e                 213.5   
43               2.47            1.25      b      d                 224.6   
54               2.27            1.25      b      d                 224.6   
103              1.64            1.25      d      d                 179.8   
2                1.00            0.50      b      b                  61.3   
3                1.00            0.50      b      b                  61.3   

     Expected_charge  Difference(Courier-Expected_X)  
102             45.4                             0.0  
56              90.2                           -44.8  
79              33.0                            28.3  
80              33.0                             0.0  
85             107.3                           106.2  
43              90.2                           134.4  
54              90.2                           134.4  
103             90.2                            89.6  
2               61.3                             0.0  
3               33.0                            28.3

In [ ]:
len(Output.columns)

11

In [ ]:
Output.columns = ["Order ID",	"AWB Number",	"Total weight as per X (KG)",	
                  "Weight slab as per X (KG)",	"Total weight as per Courier Company (KG)",	
                  "Weight slab charged by Courier Company (KG)",	"Delivery Zone as per X",	
                  "Delivery Zone charged by Courier Company",	"Expected Charge as per X (Rs.)",	
                  "Charges Billed by Courier Company (Rs.) ",	
                  "Difference Between Expected Charges and Billed Charges (Rs.)"]

In [336]:
Output.to_csv("Calculations.csv", index = False)

##Summary

In [351]:
Output_summary = pd.DataFrame(columns=['Condition', 'Count', 'Amount(Rs).', 'Overpaid'])

In [352]:
Output_summary.columns

Index(['Condition', 'Count', 'Amount(Rs).', 'Overpaid'], dtype='object')

In [353]:
Output_summary["Condition"] = ["Total orders where X has been correctly charged", "Total Orders where X has been overcharged", "Total Orders where X has been undercharged"]

In [354]:
Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged   NaN         NaN      NaN
1        Total Orders where X has been overcharged   NaN         NaN      NaN
2       Total Orders where X has been undercharged   NaN         NaN      NaN

In [355]:
count_correct = (Output["Difference(Courier-Expected_X)"] == 0).sum()

Output_summary.loc[0, "Count"] = count_correct

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41         NaN      NaN
1        Total Orders where X has been overcharged   NaN         NaN      NaN
2       Total Orders where X has been undercharged   NaN         NaN      NaN

In [356]:
count_overcharge = (Output["Difference(Courier-Expected_X)"] > 0).sum()

Output_summary.loc[1, "Count"] = count_overcharge

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41         NaN      NaN
1        Total Orders where X has been overcharged    76         NaN      NaN
2       Total Orders where X has been undercharged   NaN         NaN      NaN

In [357]:
count_undercharge = (Output["Difference(Courier-Expected_X)"] < 0).sum()

Output_summary.loc[2, "Count"] = count_undercharge

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41         NaN      NaN
1        Total Orders where X has been overcharged    76         NaN      NaN
2       Total Orders where X has been undercharged     7         NaN      NaN

In [358]:
sum_correct = Output[Output["Difference(Courier-Expected_X)"] == 0]["C_Billing_Amount(Rs."].sum()

Output_summary.loc[0, "Amount(Rs)."] = sum_correct

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      NaN
1        Total Orders where X has been overcharged    76         NaN      NaN
2       Total Orders where X has been undercharged     7         NaN      NaN

In [359]:
sum_overcharge = Output[Output["Difference(Courier-Expected_X)"] > 0]["C_Billing_Amount(Rs."].sum()

Output_summary.loc[1, "Amount(Rs)."] = sum_overcharge

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      NaN
1        Total Orders where X has been overcharged    76      9797.4      NaN
2       Total Orders where X has been undercharged     7         NaN      NaN

In [360]:
sum_undercharge = Output[Output["Difference(Courier-Expected_X)"] < 0]["C_Billing_Amount(Rs."].sum()

Output_summary.loc[2, "Amount(Rs)."] = sum_undercharge

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      NaN
1        Total Orders where X has been overcharged    76      9797.4      NaN
2       Total Orders where X has been undercharged     7       581.8      NaN

In [363]:
sum_correct = Output[Output["Difference(Courier-Expected_X)"] == 0]["Difference(Courier-Expected_X)"].sum()

Output_summary.loc[0, "Overpaid"] = sum_correct

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      0.0
1        Total Orders where X has been overcharged    76      9797.4      NaN
2       Total Orders where X has been undercharged     7       581.8      NaN

In [364]:
sum_correct = Output[Output["Difference(Courier-Expected_X)"] > 0]["Difference(Courier-Expected_X)"].sum()

Output_summary.loc[1, "Overpaid"] = sum_correct

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      0.0
1        Total Orders where X has been overcharged    76      9797.4   5289.4
2       Total Orders where X has been undercharged     7       581.8      NaN

In [365]:
sum_correct = Output[Output["Difference(Courier-Expected_X)"] < 0]["Difference(Courier-Expected_X)"].sum()

Output_summary.loc[2, "Overpaid"] = sum_correct

Output_summary.head()

Condition Count Amount(Rs). Overpaid
0  Total orders where X has been correctly charged    41      3269.0      0.0
1        Total Orders where X has been overcharged    76      9797.4   5289.4
2       Total Orders where X has been undercharged     7       581.8   -259.2

In [366]:
Output_summary.to_csv("Summary.csv", index = False)